plugin-api/src/main/java/org/hyperledger/besu/plugin/services/storage/LayeredKeyValueStorage.java

```diff
@@ -0,0 +1,141 @@
+import java.util.concurrent.locks.Lock;
+import java.util.stream.Stream;
+
+import com.google.common.collect.ImmutableSet;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.tuweni.bytes.Bytes;
+
+/** Key value storage which stores in memory all updates to a parent worldstate storage. */
+public class LayeredKeyValueStorage extends InMemoryKeyValueStorage
+    implements SnappedKeyValueStorage {
+
+  private final KeyValueStorage parent;
+
+  /**
+   * Instantiates a new Layered key value storage.
+   *
+   * @param parent the parent key value storage for this layered storage.
+   */
+  public LayeredKeyValueStorage(final KeyValueStorage parent) {
+    this(new ConcurrentHashMap<>(), parent);
+  }
+
+  /**
+   * Constructor which takes an explicit backing map for the layered key value storage.
+   *
+   * @param map the backing map
+   * @param parent the parent key value storage for this layered storage.
+   */
+  public LayeredKeyValueStorage(final Map<Bytes, byte[]> map, final KeyValueStorage parent) {
+    super(map);
+    this.parent = parent;
+  }
+
+  @Override
+  public boolean containsKey(final byte[] key) throws StorageException {
+    return get(key).isPresent();
+  }
+
+  @Override
+  public Optional<byte[]> get(final byte[] key) throws StorageException {
+    Bytes wrapKey = Bytes.wrap(key);
+    return Optional.of(wrapKey)
+        .flatMap(
+            keyWrapped ->
+                Optional.ofNullable(hashValueStore.get(keyWrapped))
+                    .or(
+                        () ->
+                            parent
+                                .get(key)
+                                .map(
+                                    bytes1 -> {
+                                      if (!(parent instanceof LayeredKeyValueStorage)) {
+                                        hashValueStore.put(wrapKey, bytes1);
+                                      }
+                                      return bytes1;
+                                    })))
+        .filter(bytes -> bytes.length > 0);
+  }
+
+  @Override
+  public Stream<Pair<byte[], byte[]>> stream() {
+    final Lock lock = rwLock.readLock();
+    lock.lock();
+    try {
+      return ImmutableSet.copyOf(hashValueStore.entrySet()).stream()
+          .filter(entry -> entry.getValue().length > 0)
+          .map(bytesEntry -> Pair.of(bytesEntry.getKey().toArrayUnsafe(), bytesEntry.getValue()));
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public Stream<byte[]> streamKeys() {
+    final Lock lock = rwLock.readLock();
+    lock.lock();
+    try {
+      return ImmutableSet.copyOf(hashValueStore.entrySet()).stream()
+          .filter(entry -> entry.getValue().length > 0)
+          .map(bytesEntry -> bytesEntry.getKey().toArrayUnsafe());
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public boolean tryDelete(final byte[] key) {
+    // TODO: can we rely on zero byte array to indicate deletion?
```

Comment:
@matkt I believe this is "yes", are there cases where we would legitimately want to add and store a zero byte array in key value storage rather than deleting it?

Reply:
yes because otherwise we don't keep the info and we will get it from the parent that still have it. we must mark it as delete to avoid this